bookdown/03-optimization-searchspace.Rmd

```diff
@@ -0,0 +1,231 @@
+
+Note, that `r ref("ParamSet")` objects exist in two contexts.
+First, `r ref("ParamSet")`-objects are used to define the space of valid parameter setting for a learner (and other objects).
+Second, they are used to define a search space for tuning.
+We are mainly interested in the latter.
+For an example we can consider the `minsplit` parameter of the `r ref("mlr_learners_classif.rpart", "classif.rpart Learner")`.
+The `r ref("ParamSet")` associated with the learner has a lower but *no* upper bound.
+However, for tuning the value, a lower *and* upper bound must be given because tuning search spaces need to be bounded. 
+For `r ref("Learner")` or `r ref("PipeOp")` objects, typically "unbounded" `r ref("ParamSet", text = "ParamSets")` are used.
+Here, however, we will mainly focus on creating "bounded" `r ref("ParamSet", text = "ParamSets")` that can be used for tuning. 
+See the [in-depth `paradox` chapter](#paradox) for more details on using `r ref("ParamSet", text = "ParamSets")` to  define parameter ranges for use-cases besides tuning.
+
+### Creating `ParamSet`s
+
+An empty `r ref("ParamSet")` -- not yet very useful -- can be constructed using just the `r ref("ps")` call:
+```{r 03-optimization-searchspace-000}
+library("paradox")
+search_space = ps()
+print(search_space)
+```
+
+`r ref("ps")` takes named `r ref("Domain")` arguments that are turned into parameters. A possible search space for the `"classif.svm"` learner could for example be:
+
+```{r 03-optimization-searchspace-001}
+search_space = ps(
+  cost = p_dbl(lower = 0.1, upper = 10),
+  kernel = p_fct(levels = c("polynomial", "radial"))
+)
+print(search_space)
+```
+
+There are five domain constructors that produce a parameters when given to `r ref("ps")`:
+
+| Constructor               | Description                 | Is bounded?                | Underlying Class           |
+| :-----------------------: | :-------------------------: | :------------------------: | :------------------------: |
+| `p_dbl`                   | Real valued parameter ("double") | When `upper` and `lower` are given | `r ref("ParamDbl")` |
+| `p_int`                   | Integer parameter           | When `upper` and `lower` are given | `r ref("ParamInt")`    |
+| `p_fct`                   | Discrete valued parameter ("factor") | Always            | `r ref("ParamFct")` |
+| `p_lgl`                   | Logical / Boolean parameter | Always                     | `r ref("ParamLgl")`    |
+| `p_uty`                   | Untyped parameter           | Never                      | `r ref("ParamUty")`    |
```

Comment:
Can this table to formatted so that column widths match their contents? Many editors have auto-formatters for this.